tests/integration_tests/data_steward/cdr_cleaner/cleaning_rules/bigquery_tests_base.py

```diff
@@ -0,0 +1,243 @@
+            self.assertIsNone(response.exception())
+
+        def tearDown(self):
+            """
+            Ensure the data is dropped from the table(s).
+            """
+            for table in self.fq_table_names + self.fq_sandbox_table_names:
+                self.drop_rows(table)
+
+        def load_test_data(self, sql_statements=None):
+            """
+            Add data to the tables for the rule to run on.
+
+            :param sql_statements: a list of sql statements to load for testing.
+            """
+
+            if not sql_statements or not isinstance(sql_statements, list):
+                raise RuntimeError("Provide parameter sql_statements.  The "
+                                   "parameter must be list of sql strings.")
+
+            for query in sql_statements:
+                response = self.client.query(query)
+                self.assertIsNotNone(response.result())
+                self.assertIsNone(response.exception())
+
+    class DropRowsTestBase(BigQueryTestBase):
+        """
+        Class that can be extended and used to test cleaning rules that drop row data.
+
+        This class defines a basic test that can be used by extending classes
+        in lieu of writing the test function.  The test function can be overridden
+        by extending classes, if desired.  This test will minimally ensure
+        that sandbox tables are empty and all data is loaded prior to test
+        execution.  It will then ensure only expected cleaned data exists in
+        the cleaned tables, and defined sandboxed data exists in the sandbox
+        table(s).  This class is optional and is not required.  It is here to
+        make testing easier.  All assertions are based on {tablename}_id fields.
+        """
+
+        @classmethod
+        def initialize_class_vars(cls):
+            super().initialize_class_vars()
+            # a list of dictionaries where each dictionary defines table
+            # expectations for each table to validate with rule execution, such as:
+            # 'name':  common name of the table being cleaned and defining
+            #          execution expectations for
+            # 'fq_table_name':  the fully qualified name of the table being cleaned
+            # 'fq_sanbox_table_name':  the fully qualified name of the sandbox
+            #                          table the rule will create if one exists
+            # 'loaded_ids':  The list of ids loaded by the sql insert statement
+            # 'sandboxed_ids':  the list of ids that will be in the sandbox if
+            #                   the rule sandboxes information
+            # 'cleaned_ids':  the list of ids that will continue to exist in the
+            #                 input table after running the cleaning rule
+            cls.tables_and_test_values = []
+            # The query class that is being executed.
+            cls.query_class = None
+
+        def test_execution_default(self):
```

Comment:
I was unable to reply to this comment presumably due to rebases, quoting here:

> calbach 4 days ago Member
This feels like it's forcing a single test case for the test class, which might be overly restrictive. I might just make this whole thing a helper method which would be invoked by a more specifically name subclass test, e.g. rename this to run_test_execution_of_get_query_specs({... expectations ... }) and move the test output expectations into the parameter rather than on the base class definition

> @lrwb-aou
lrwb-aou 2 hours ago • 
 Author Member
No. You're not thinking about this from an OOP perspective. I will rename the method to test_execution_default, but I still think this should be a test because it is a default method. Meaning, if you don't declare any tests, this will execute for a properly set up extending class. If you don't like what this method is doing, you can override this method in your extending class with a new definition (even if it's just a pass body. But at that point, I probably wouldn't extend this class). If you like what this method is doing and want to add additional tests, you can define the additional tests in your extending class.

I understand what you did and like to think I have a pretty reasonable grasp of OOP. My comment still stands. In my experience putting test cases in base classes like this can be pretty convoluted to trace, and my initial impression is that a reasonably comprehensive test will want to run more than one scenario. Test naming and tracing also don't play well with this model. Adding extra tests to a subclass is a bit awkward, since it would mean the above test setup (done for the whole unit test) only applies to a single test case.

If you think the ideal end state is a large number of child test cases which simply execute the default behavior here, then this pattern is reasonable to save boilerplate. I'm skeptical this is sufficient but if you feel strongly I'll be happy to be proven wrong. Please give it some consideration - but fine to merge as-is.

Reply:
Ideally, yes, the test author would provide the information and set up their test with enough realistic test data to cover all the possible scenarios.  The idea was to save a bunch of boiler plate and ensure the test data loaded correctly.  This would ensure a minimal level of testing which we're not getting right now, hopefully with as little effort as possible.

I have refactored and abandoned the idea of a drop rows base test class and modify rows base test class.  I will instead add to a cleaning rules base test class in the future another method that must be specially invoked by extending subclasses.  I doubt they will be used much because they have to be specifically invoked.